json-ld/src/main/scala/io/renku/jsonld/Flattener.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 Swiss Data Science Center (SDSC)
+ * A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+ * Eidgenössische Technische Hochschule Zürich (ETHZ).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.renku.jsonld
+
+import cats.syntax.all._
+import io.renku.jsonld.JsonLD.{JsonLDArray, JsonLDEntity, JsonLDEntityId, MalformedJsonLD}
+
+import scala.annotation.tailrec
+
+trait JsonLDArrayFlattener extends Flattener {
+  self: JsonLDArray =>
+  lazy val flatten: Either[MalformedJsonLD, JsonLD] =
+    for {
+      flattenedJsons <- this.jsons
+                          .foldLeft(Either.right[MalformedJsonLD, List[JsonLD]](List.empty[JsonLD])) {
+                            case (acc, jsonLDEntity: JsonLDEntity) =>
+                              for {
+                                jsons    <- deNest(List(jsonLDEntity), List.empty[JsonLDEntity])
+                                accRight <- acc
+                              } yield accRight ++ jsons
+                            case (acc, other) => acc.map(other +: _)
+                          }
+      flattenedArray <- checkForUniqueIds(flattenedJsons.distinct)
+    } yield flattenedArray
+}
+
+trait JsonLDEntityFlattener extends Flattener {
+  self: JsonLDEntity =>
+
+  lazy val flatten: Either[MalformedJsonLD, JsonLD] = deNest(List(this), Nil).flatMap(checkForUniqueIds)
+}
+
+private[jsonld] trait Flattener {
```

Comment:
What would you think if we create a new 'flatten' package for these three traits? Then, effectively we could split this file into three (or four) files for each trait.

I think there could be some advantages of that:
* smaller compilation units;
* the `deNest` method could be `protected[flatten]` scope;
* the `checkForUniqueIds ` method could be `protected[flatten]` scope;

By the way, maybe the `checkForUniqueIds` could go into a separate trait like `Validation` in the `flatten` package?